src/backend/backend.c

```diff
@@ -204,15 +204,17 @@ void paint_all_new(session_t *ps, struct managed_win *t, bool ignore_damage) {
 			if (w->state == WSTATE_MAPPING) {
 				// Gradually increase the blur intensity during
 				// fading in.
-				blur_opacity = w->opacity / w->opacity_target;
```

Comment:
The intention here is that as `w->opacity` gradually goes from `0` to `w->opacity_target`, `blur_opacity` gradually increase to 1. I don't understand what using multiply here means.

Reply:
`w->opacity` should not exceed `w->opacity_target`, if it does, there is a bug somewhere else in my code.

Reply:
Ah, I didn't read the comments correctly, I thought that code was meant to *decrease* the blur opacity. I think the real problem is that `opacity_target` exceeds `opacity`, then, yes.

Reply:
`Assertion `w->opacity <= w->opacity_target' failed.`

So yes, that is the problem. It takes some effort to trigger it, though... Maybe a floating point error?

Reply:
So perhaps a hint:

```
[ 02/23/2020 22:45:41.555 paint_all_new TRACE ] 0.870000 <= 1.000000
BLAH BLAH BLAH
[ 02/23/2020 22:45:41.573 paint_all_new TRACE ] 0.810000 <= 0.625000

picom: ../src/backend/backend.c:208: paint_all_new: Assertion `w->opacity <= w->opacity_target' failed.
```

It looks like `opacity_target` is changed while `opacity` is still lower than the new value.

Reply:
Ah, I think I understand the problem...

```
$ cat ~/.xmonad/xmonad.hs | grep 0.375
        [ isUnfocused --> transparency 0.375
```

`0.625` is the default opacity of unfocused windows on my setup. It looks like picom incorrectly assumes the opacity of an unmapped/mapping window can never change.

As for a quick solution:

I think it is reasonable to just jump to the `opacity_target` when we detect an external program changed the opacity during a mapping. This should be rare, and is probably the 'intended' opacity in that case.

Reply:
This doesn't solve the problem for unmapping windows, though. Also, I think the problem is really somewhere else in the code

Reply:
Also, while I was testing, I got this:
```
Assertion `(w->flags & WIN_FLAGS_SHADOW_STALE) == 0' failed.
```

Yeah, I'm not getting into that right now. I'll save the log, though

Reply:
> It looks like picom incorrectly assumes the opacity of an unmapped/mapping window can never change.

This is a problem as it certainly can change by changing focus (`inactive-opacity`) or if set directly, e.g. via `picom-trans`.

> As for a quick solution:
> 
> I think it is reasonable to just jump to the `opacity_target` when we detect an external program changed the opacity during a mapping. This should be rare, and is probably the 'intended' opacity in that case.

`blur_opacity` is intended to be 1, if the window is `MAPPED` (and visible) and not directly dependent on window opacity EXCEPT when fading in/out. So clamping to 1 should be more reasonable.